### ISSUE TITLE
Google V3 Mapstraction behavior

### ISSUE THREAD

**Issue by mjesun** [mapstraction/mxn#223]
Hi all,

I've recently started using Mapstraction at work, so I'm not quite familiar with the interface, but I think I've found what could be considered a bad behavior.

We are currently using Google Maps V3, putting many markers over a map, and adding an infoWindow to each of them. As you know, V3 allows you to have many infowindows shown at the same time, but that is not a behavior we want to have, so I wrote a code like this:

``` js
function openLayer(event, marker) {
    map.markers.forEach(function(marker) {
        marker.closeBubble();
    });

    marker.openBubble();
}
```

Testing this code, I found that, on the first click, two layers are created for the same marker, thus making impossible to close the first one. After digging in the code, I found the following snippet:

``` js
if (this.infoBubble) {
    var event_action = "click";
    if (this.hover) {
        event_action = "mouseover";
    }
    google.maps.event.addListener(marker, event_action, function() {
        marker.mapstraction_marker.openBubble();
    });
}
```

Which is automatically adding a listener for showing the layer. If I comment the handler, everything works fine. I could not trace why the info window is created twice.

But, apart from this, should mapstraction add default handlers? I think it would be better if this is explicitly done rather than implicitly.

**Comment by vicchi** [comment]
Hi ... thanks for this. Are you using the `master` branch of Mapstraction? If so, could you re-test against the `release-2.1` branch and let me know if this is still an issue?

If it is, could you post a working example, preferably as a Gist, so I can take a better look at this?

-Gary

**Comment by mjesun** [comment]
Hi,

We've reproduced the same behavior with the `release-2.1` branch, so we isolated our code and created a gist with it: https://gist.github.com/mjesun/5168429. Let me know if you need more information.

**Comment by vicchi** [comment]
Thanks. I'll take a look later today or over the weekend.

**Comment by vicchi** [comment]
There's something very odd going on here but I haven't yet worked out what it is. It seems the order of Marker creation is important. If I use your test code and click on the left most marker and then on the right, both infoBubbles stay open. If I then close down the infoBubbles and click on the right most marker and then the left most marker, it works. It's also the same if I refresh the browser, right most and then left most works. This is bizarre. Will keep looking at this.

**Comment by mjesun** [comment]
Hi,

Well, technically I've found that the problem is not related to closing the window, but to two infoBubbles superimposed one on top of the other. I noticed that by looking to the drop shadow that Google generates for each infoBubble (open the frist bubble, then try closing it; you'll see that the infoBubble still exists, but the shadow englighten).

**Comment by vicchi** [comment]
I can see that the infoBubble is creating the shadow, as it should, but I think this may be down to the fact that the event handlers Mapstraction enqueues on your behalf, via the native Google API event system, fire first, before the Mapstraction events fire. But that still doesn't explain why a native Google API call to close the infoBubble isn't working in one order but does in the other. I'll keep looking ...

**Comment by vicchi** [comment]
And also, I don't see two infoBubble layers being created for a single Marker, which is what I think you were saying in your initial post?

**Comment by vicchi** [comment]
One other thing. I agree that the default click/hover handlers get in the way with a use case like this. I'll look to making them the default but to allow an optional override for custom case like the one you have. That way your use case will be supported but we won't regress and break existing code.

**Comment by mjesun** [comment]
I've attached some images so you can see where are the two infoBubble layers:

First, as you said, you need to click on the left marker:
![step_1](https://f.cloud.github.com/assets/829269/308911/6a431528-96f4-11e2-9101-a79c2bdfae2d.png)

Then, an (apparently) single infoBubble will appear:
![step_2](https://f.cloud.github.com/assets/829269/308912/6a577ef0-96f4-11e2-9b2c-f6ebd8f68533.png)

But if you try to close it, you'll see that, in fact, there were two infoBubbles, one on top of the other.
![step_3](https://f.cloud.github.com/assets/829269/308913/6a7097dc-96f4-11e2-8299-7d1e767de78b.png)

As you can see, the shadow enlightened, because there was a shadow on top of a shadow. Now, if you close that second bubble, it will successfully close. You can corroborate the fact that there are two infoBubble layers by looking at the generated DOM (inside the Google map).

**Comment by vicchi** [comment]
Now I see _exactly_ what you mean but I still don't understand why this is happening. I'll keep on digging into this.

**Comment by vicchi** [comment]
Some progress. It's definitely something to do with the default click handler that `mxn.Marker.toProprietary` installs when a Marker is created if there's an `infoBubble` specified.

If I change your test code slightly so that the markers are added to the map _before_ the call to `mxn.Marker.setInfoBubble()` is made, then the default handlers aren't installed (which you could argue is a bug in itself, or at least inconsistent behaviour), then your test code works as you would think it should. In other words, rather than having this in your test code ...

``` javascript
marker1.setInfoBubble('marker1');
map.addMarker(marker1);

marker2.setInfoBubble('marker2');
map.addMarker(marker2);
```

... I now have this ...

``` javascript
map.addMarker(marker1);
map.addMarker(marker2);

marker1.setInfoBubble('marker1');
marker2.setInfoBubble('marker2');
```

Which points the finger firmly at the behaviour of the default click/hover handler that `mxn.Marker.toProprietary` installs. What I don't yet understand is _why_ this happens. But it's some progress and has narrowed the parameters for tracking this down somewhat.

**Comment by mjesun** [comment]
I think I have narrowed even more the problem. After your discoveries, I found that the problem is related to calling `closeBubble` over the marker which is going to be opened, _before_ it actually opens. This makes some strange behavior involving the layer opened on the default click handler, which makes in fact opening two layers.

I found this by blocking the first call to closeBubble, by using this code snippet:

``` js
var notFirstOpen = false;

var handler = function(e, marker) {
    if (notFirstOpen) {
        //We close alll the existing bubbles.
        map.markers.forEach(function(item, i) {
            item.closeBubble();
        });
    }

    notFirstOpen = true;

    //We open the corresponding bubble.
    marker.openBubble();
};
```

**Comment by vicchi** [comment]
That makes more sense. I think the way forward is to make the default handlers an option, as I mentioned earlier, so that existing code doesn't break, but you can override the default click/hover being implemented if you want to handle the click/hover event yourself. I haven't written the code to do this yet but I image it would look something like this ...

``` javascript
var marker1 = new mxn.Marker(latLon1);
marker1.setInfoBubble('marker1');
marker1.setOptions({ event: 'none' });
```

... where the options passed into the new `mxn.Marker.setOptions()` call would be one of `click`, `hover` or `none`. `click` and `hover` mirror the current implementation whilst `none` disables the installing of an `infoBubble` event handler entirely.

**Comment by gilesc50** [comment]
+1 Vote for this idea.

Giles

**Comment by vicchi** [comment]
I can see how the Easter weekend is shaping up ....

**Comment by mjesun** [comment]
I agree with the idea of controlling the default handlers; nevertheless, I'm still worried about why when changing the call order (as you pointed [before](https://github.com/mapstraction/mxn/issues/223#issuecomment-15578968)) solves the original issue.

So, apart from letting the user to control whether the default handlers can be disabled or not (through `setOptions`), I think it would be great to also deactivate this behavior whenever you attach one `click` handler. That would not break the existing code too, and it would be even more intuitive.

What do you think?

**Comment by vicchi** [comment]
I think I agree. Just to confirm, you mean that this code ..

``` javascript
var marker = new mxn.Marker(coords);
```

... would automagically install the click/hover handler (unless you've disabled this behaviour via options). If I then did this ...

``` javascript
marker.click.addHandler(handler);
```

... would unregister the default click/hover handler and make `handler` the sole event handler for the `click` event for that specific marker.

Is that your understanding too?

**Comment by mjesun** [comment]
Exactly. I think is the best approach, so people already having their own handlers (and possibly experiencing this issue) will get the bug corrected without changing their code.

**Comment by vicchi** [comment]
I think we have a deal. I'll get this implemented and pushed to the `release-2.1` branch as soon as I can, which will probably be over the weekend or early next week.

**Comment by mjesun** [comment]
Sounds really great! :) Let me know then if you need anything else.

**Comment by vicchi** [comment]
Tentative progress. I'm putting this together on a new branch ... `release-2.1-marker-events`. I've changed `mxn.core.js` to add a new Marker attribute, `defaultHandler`, which is supported via `Marker.addData()` and the (new) `Marker.setDefaultHandler()` methods.

Take a look at `tests/marker-test.html` (https://github.com/mapstraction/mxn/blob/release-2.1-marker-events/tests/marker-test.html); it's a reworking of your original Gist test case, which now works the way it should. I think.

Let me know what you think ...

**Comment by mjesun** [comment]
I've been looking the test, and that seems OK for me, though I think that the `setDefaultHandler(false)` call should be removable ([this one](https://github.com/mapstraction/mxn/blob/b2f881e8c73456dea061d13180cc3dfe57b5eb25/tests/marker-test.html#L39)). For me, when making the `click.addEvent(...)` call, the default handler should be detached; but as I've seen so far this is not so easy to implement, because there is no way of knowing when a event has been added.

From that point of view, I can see three posible solutions:
- First, add [here](https://github.com/mapstraction/mxn/blob/b2f881e8c73456dea061d13180cc3dfe57b5eb25/source/mxn.js#L172) a trigger for a general event like `eventAdded` or something like that. This can become dangerous if somebody tries to attach an event within `eventAdded` is triggered (you would run into recursive calls until a `StackOverflow` was thrown).
- Or, you could overwrite, only for `mxn.Marker`, the functions `addHandler` and `removeHandler`; which is not an elegant solution.
- Or, we could left as-is now (I mean with the current fix); although maybe the incidence will be (erroneously then) reported again by someone.

**Comment by vicchi** [comment]
This is still very much a work in progress. I've put together a first cut of stopping the default events being installed. Which seems to work as intended. I still want to deregister the default handlers in a sane way when a new click handler or mouseover handler is installed; that's the next cut of the code. Quite how this will happen, if at all, remains to be seen, for all the reasons you've stated above (and probably more).

We'll have to see how this turns out. I don't want to leave this as solution three (leave as-is) but at least there's a fallback position if all other approaches don't really work.

More coding needed.

**Comment by vicchi** [comment]
Actually this may be easier than I first thought (famous last words).  When instantiating a new `mxn.Marker`, `mxn.AddEvents` is called to add support for the `openInfoBubble`, `closeInfoBubble` and `click` events. `mxn.AddEvents` instantiates a new `mxn.Event` for each event, passing the parent `mxn.Marker` object as a parameter. But `mxn.Event` never stores or uses the parent object. If I extend the base `mxn.Event` class, I can determine whether this is a `mxn.Marker` event being fired.

The other change is that the default events are proprietary and the event handlers are closures. If I store the return from the proprietary event handler registration (which works with Google v3) then I should be able to uninstall the default handlers when other handlers are installed. I'll have to look and see how well this approach will work with other providers, but it seems a way forward.

Still more coding needed but at least there's a germ of a solution in here somewhere.

Note to self: do we want to extend the `mxn.Marker` events to include `mouseover` and `mouseout`?

Other note to self: if this gets solved for `mxn.Marker` it seems only logical to extend this approach to `mxn.Polyline` as well.

**Comment by mjesun** [comment]
Glad to hear that you mantain the original idea of unregistering the default handler :)

Regarding the extension of `mxn.Event` class, that would mean that you would need another additional "`addEvents`" method for instancing those particular event objects inside the `mxn.Marker` constructor; which I think it would add complexity to the existing code. In my opinion, the code for detecting when someone attaches a listener, should be located on the base class; but I it's true that I can't realize a good way for notifying the attaching itself.

For unregistering the event it is pretty likely to run into trouble to store the result of `addListener` inside a variable for other map providers; in fact, is the first time that I see that a method for attaching handlers works this way; they usually remove the handler by passing the original function you attached (as the DOM does, and as you do in Mapstraction too).

And, yep, I think it's a great idea to have control over `mouseover` and `mouseout`. If you have a `hover` property (toggable through `setHover`) for choosing whether you want to show the infoBubble on a `hover` or a `click`, it seems quite natural to support those events too. But we'd separate this into another issue :)

**Comment by mjesun** [comment]
I have just noticed the issue #144, which affects to a Marker, and only with V3. The latest comment says that it's related to `toPropietary` method, so it will probably has the same origin than this one. I'll investigate further.

**Comment by vicchi** [comment]
I've just taken a step back for a moment and looked at the other provider implementations and noticed something which I think we need to take into consideration.

Looking at the rest of the Mapstraction code, this explicit default handler code only appears in the Google v2, Google v3, Ovi and Nokia provider implementations. Which actually means, in a slightly roundabout way, that it's only implemented in the Google provider code as I wrote both the Ovi and Nokia implementations and based them on the Google v3 code as a reference implementation.

The other providers support this default handler in a variety of ways ... Leaflet for `click` but not `mouseover` ... OpenLayers not at all.

We really want to maintain parity across all the providers, wherever possible. This is the core feature of Mapstraction and should really underpin the entire code base.

If we want to maintain parity, we really need to implement the `Marker` default handler across _all_ providers.

So even when we get this sorted for Google v3, there's all the other providers to look at too.

More coding.

**Comment by vicchi** [comment]
See also issues #239 and the related #238, which I've just opened.

**Comment by riklaunim** [comment]
I'm using the dev "v3" branch and in openBubble for googlev3 I've added:

```
        clearExistingInfoBubbles(this.mapstraction.markers);
        function clearExistingInfoBubbles(markers) {
            markers.forEach(function(marker) {
                marker.closeBubble();
            });
        }
```

Which closes open infoBubbles when another marker is clicked. More generic approach would be to allow passing callbacks to methods like open/close bubble. In this case maybe even before/after callbacks.
